op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -421,37 +454,54 @@ func runStateMigration(newDBPath string, opts stateMigrationOptions) error {
 	}
 	config.SetDeployments(deployments)
 
-	// Get latest block information from L1
 	var l1StartBlock *types.Block
 	client, err := ethclient.Dial(opts.l1RPC)
 	if err != nil {
 		return fmt.Errorf("cannot dial %s: %w", opts.l1RPC, err)
 	}
-
+	// If the L1 starting block tag is not set, we determine it dynamically by
+	// finding the most recent final L1 block at the time of the L2 fork block.
 	if config.L1StartingBlockTag == nil {
-		l1StartBlock, err = client.BlockByNumber(context.Background(), nil)
+		// Find the L1 starting block, the L2 fork block occurs 1 minute after the last celo L1 block.
+		opts.migrationBlockTime = celoL1Head.Time + 60
+		// Wait for the L2 start time before we calculate the L1 starting block.
+		tillL2Start := int64(opts.migrationBlockTime) - time.Now().Unix()
+		if tillL2Start >= 0 {
+			time.Sleep(time.Duration(tillL2Start+1) * time.Second)
+		}
+		l1StartingBlockHash, err := NewBeaconClient(opts.l1BeaconRPC, opts.l1BeaconchainURL).MostRecentFinalizedL1BlockAtTime(opts.migrationBlockTime)
 		if err != nil {
-			return fmt.Errorf("cannot fetch latest block: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock based on L2 starting block time time (%v): %w", opts.migrationBlockTime, err)
 		}
-		tag := rpc.BlockNumberOrHashWithHash(l1StartBlock.Hash(), true)
-		config.L1StartingBlockTag = (*genesis.MarshalableRPCBlockNumberOrHash)(&tag)
-	} else if config.L1StartingBlockTag.BlockHash != nil {
+		config.L1StartingBlockTag = &genesis.MarshalableRPCBlockNumberOrHash{BlockHash: &l1StartingBlockHash}
+	}
+
+	if config.L1StartingBlockTag.BlockHash != nil {
 		l1StartBlock, err = client.BlockByHash(context.Background(), *config.L1StartingBlockTag.BlockHash)
 		if err != nil {
-			return fmt.Errorf("cannot fetch block by hash: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock by hash (%v): %w", config.L1StartingBlockTag.BlockHash, err)
 		}
 	} else if config.L1StartingBlockTag.BlockNumber != nil {
 		l1StartBlock, err = client.BlockByNumber(context.Background(), big.NewInt(config.L1StartingBlockTag.BlockNumber.Int64()))
 		if err != nil {
-			return fmt.Errorf("cannot fetch block by number: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock by number (%v): %w", config.L1StartingBlockTag.BlockNumber, err)
 		}
 	}
 
-	// Ensure that there is a starting L1 block
-	if l1StartBlock == nil {
-		return fmt.Errorf("no starting L1 block")
+	// We want to ensure that we have sufficient time to start the sequencer
+	// before it exceeds max sequencer drift which is currently 2892s ==
+	// 48.2mins. Devops need about 7 mins to get the sequencer started after the
+	// migration script is run. We give them a 10 minute buffer in case of any
+	// issues, but if we are outside of this range we will fail here.
+	var maxSequencerDriftCelo uint64 = 2892 // Duplicate of the value defined at op-node/rollup.maxSequencerDriftCelo
```

Comment:
Why not use the value from there?

Reply:
Its private and meant to be accessed via the ChainSpec object, so exporting the plain value might encourage incorrect usage, also if we try to re-run the migration at a later date we want this value to be the same, even if it changed elsewhere in the code. So it seemed like a safer option to just duplicate the value here, with the assumption that it's pretty unlikely to change before we launch mainnet.